File: samples/java/youtube-cmdline-addfeaturedvideo-sample/src/main/java/com/google/api/services/samples/youtube/cmdline/youtube_cmdline_addfeaturedvideo_sample/AddFeaturedVideo.java
/*
 * Copyright (c) 2013 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

package com.google.api.services.samples.youtube.cmdline.youtube_cmdline_addfeaturedvideo_sample;

import com.google.api.client.auth.oauth2.Credential;
import com.google.api.client.extensions.java6.auth.oauth2.AuthorizationCodeInstalledApp;
import com.google.api.client.extensions.java6.auth.oauth2.FileCredentialStore;
import com.google.api.client.extensions.jetty.auth.oauth2.LocalServerReceiver;
import com.google.api.client.googleapis.auth.oauth2.GoogleAuthorizationCodeFlow;
import com.google.api.client.googleapis.auth.oauth2.GoogleClientSecrets;
import com.google.api.client.googleapis.json.GoogleJsonResponseException;
import com.google.api.client.http.HttpTransport;
import com.google.api.client.http.javanet.NetHttpTransport;
import com.google.api.client.json.JsonFactory;
import com.google.api.client.json.jackson2.JacksonFactory;
import com.google.api.services.youtube.YouTube;
import com.google.api.services.youtube.model.*;
import com.google.common.collect.Lists;

import java.io.File;
import java.io.IOException;
import java.math.BigInteger;
import java.util.List;

/**
 * This program adds a featured video to a channel via the Invideo Programming API.
 *
 * @author Ikai Lan <ikai@google.com>
 */
public class AddFeaturedVideo {

    /**
     * Global instance of the HTTP transport.
     */
    private static final HttpTransport HTTP_TRANSPORT = new NetHttpTransport();

    /**
     * Global instance of the JSON factory.
     */
    private static final JsonFactory JSON_FACTORY = new JacksonFactory();

    /**
     * Global instance of Youtube object to make all API requests.
     */
    private static YouTube youtube;


    /**
     * Authorizes the installed application to access user's protected data.
     *
     * @param scopes list of scopes needed to run youtube upload.
     */
    private static Credential authorize(List<String> scopes) throws IOException {

        // Load client secrets.
        GoogleClientSecrets clientSecrets = GoogleClientSecrets.load(
                JSON_FACTORY, AddFeaturedVideo.class.getResourceAsStream("/client_secrets.json"));

        // Checks that the defaults have been replaced (Default = "Enter X here").
        if (clientSecrets.getDetails().getClientId().startsWith("Enter")
                || clientSecrets.getDetails().getClientSecret().startsWith("Enter ")) {
            System.out.println(
                    "Enter Client ID and Secret from https://code.google.com/apis/console/?api=youtube"
                            + "into youtube-cmdline-addfeaturedvideo-sample/src/main/resources/client_secrets.json");
            System.exit(1);
        }

        // Set up file credential store.
        FileCredentialStore credentialStore = new FileCredentialStore(
                new File(System.getProperty("user.home"), ".credentials/youtube-api-addfeaturedvideo.json"),
                JSON_FACTORY);

        // Set up authorization code flow.
        GoogleAuthorizationCodeFlow flow = new GoogleAuthorizationCodeFlow.Builder(
                HTTP_TRANSPORT, JSON_FACTORY, clientSecrets, scopes).setCredentialStore(credentialStore)
                .build();

        // Build the local server and bind it to port 8080
        LocalServerReceiver localReceiver = new LocalServerReceiver.Builder().setPort(8080).build();

        // Authorize.
        return new AuthorizationCodeInstalledApp(flow, localReceiver).authorize("user");
    }

    /**
     * This is a very simple code sample that looks up a user's channel, then features the most recently
     * uploaded video in the bottom left hand corner of every single video in the channel.
     *
     * @param args command line args (not used).
     */
    public static void main(String[] args) {

        // An OAuth 2 access scope that allows for full read/write access.
        List<String> scopes = Lists.newArrayList("https://www.googleapis.com/auth/youtube");

        try {
            // Authorization.
            Credential credential = authorize(scopes);

            // YouTube object used to make all API requests.
            youtube = new YouTube.Builder(HTTP_TRANSPORT, JSON_FACTORY, credential).setApplicationName(
                    "youtube-cmdline-addfeaturedvideo-sample").build();

            // Fetch the user's channel. We also fetch the uploads playlist so we can use this later
            // to find the most recently uploaded video
            ChannelListResponse channelListResponse = youtube.channels().list("id,contentDetails")
                    .setMine(true)
                    .setFields("items(contentDetails/relatedPlaylists/uploads,id)")
                    .execute();

            // This assumes the user has a channel already. If the user does not have a channel, this should
            // throw a GoogleJsonResponseException explaining the issue
            Channel myChannel = channelListResponse.getItems().get(0);
            String channelId = myChannel.getId();
            String uploadsPlaylistId = myChannel.getContentDetails().getRelatedPlaylists().getUploads();

            // Fetch the most recently uploaded video
            PlaylistItemListResponse playlistItemListResponse = youtube.playlistItems().list("snippet")
                    .setPlaylistId(uploadsPlaylistId)
                    .setFields("items/snippet")
                    .execute();

            String featuredVideoId;
            if (playlistItemListResponse.getItems().isEmpty()) {
                // There are no videos on the channel. Therefore, we cannot feature a video. Exit.
                System.out.println("Channel contains no videos. Featuring a default video instead from the Google Developers channel.");
                featuredVideoId = "w4eiUiauo2w";
            } else {
                // The latest video should be the first video in the playlist response
                PlaylistItem featuredVideo = playlistItemListResponse.getItems().get(0);
                featuredVideoId = featuredVideo.getSnippet()
                        .getResourceId()
                        .getVideoId();

                System.out.println("Featuring video: " + featuredVideo.getSnippet().getTitle());
            }

            // Feature this video on the channel via the Invideo programming API
            // This describes the position of the video. Valid positions are bottomLeft, bottomRight, topLeft and
            // topRight
            InvideoPosition invideoPosition = new InvideoPosition();
            invideoPosition.setCornerPosition("bottomLeft");
            invideoPosition.setType("corner");

            // The allowed offsets are offsetFromEnd and offsetFromStart, with offsetMs being an offset in milliseconds
            InvideoTiming invideoTiming = new InvideoTiming();
            invideoTiming.setOffsetMs(BigInteger.valueOf(15000l));
            invideoTiming.setType("offsetFromEnd");

            // Represents the type of promotion. In this case, a video with a video ID
            PromotedItemId promotedItemId = new PromotedItemId();
            promotedItemId.setType("video");
            promotedItemId.setVideoId(featuredVideoId);

            // Construct the Invidideo promotion
            InvideoPromotion invideoPromotion = new InvideoPromotion();
            invideoPromotion.setPosition(invideoPosition);
            invideoPromotion.setTiming(invideoTiming);
            invideoPromotion.setItems(Lists.newArrayList(promotedItemId));

            // Now let's add the invideo promotion to the channel
            Channel channel = new Channel();
            channel.setId(channelId);
            channel.setInvideoPromotion(invideoPromotion);

            // Make the API call
            Channel updateChannelResponse = youtube.channels()
                    .update("invideoPromotion", channel)
                    .execute();

            // Print out returned results.
            System.out.println("\n================== Updated Channel Information ==================\n");
            System.out.println("\t- Channel ID: " + updateChannelResponse.getId());

            InvideoPromotion promotion = updateChannelResponse.getInvideoPromotion();
            System.out.println("\t- Invideo promotion video ID: " + promotion.getItems()
                    .get(0)
                    .getVideoId());
            System.out.println("\t- Promotion position: " + promotion.getPosition().getCornerPosition());
            System.out.println("\t- Promotion timing: " + promotion.getTiming().getOffsetMs()
                    + " Offset: " + promotion.getTiming().getType());
        } catch (GoogleJsonResponseException e) {
            System.err.println("GoogleJsonResponseException code: " + e.getDetails().getCode() + " : "
                    + e.getDetails().getMessage());
            e.printStackTrace();

        } catch (IOException e) {
            System.err.println("IOException: " + e.getMessage());
            e.printStackTrace();
        }
    }

}

File: samples/python/add_featured_video.py
#!/usr/bin/python

import httplib2
import os
import sys

from apiclient.discovery import build
from oauth2client.file import Storage
from oauth2client.client import flow_from_clientsecrets
from oauth2client.tools import run

from optparse import OptionParser

# CLIENT_SECRETS_FILE, name of a file containing the OAuth 2.0 information for
# this application, including client_id and client_secret. You can acquire an
# ID/secret pair from the API Access tab on the Google APIs Console
#   http://code.google.com/apis/console#access
# For more information about using OAuth2 to access Google APIs, please visit:
#   https://developers.google.com/accounts/docs/OAuth2
# For more information about the client_secrets.json file format, please visit:
#   https://developers.google.com/api-client-library/python/guide/aaa_client_secrets
# Please ensure that you have enabled the YouTube Data API for your project.
CLIENT_SECRETS_FILE = "client_secrets.json"

# Helpful message to display if the CLIENT_SECRETS_FILE is missing.
MISSING_CLIENT_SECRETS_MESSAGE = """
WARNING: Please configure OAuth 2.0

To make this sample run you will need to populate the client_secrets.json file
found at:

   %s

with information from the APIs Console
https://code.google.com/apis/console#access

For more information about the client_secrets.json file format, please visit:
https://developers.google.com/api-client-library/python/guide/aaa_client_secrets
""" % os.path.abspath(os.path.join(os.path.dirname(__file__),
                                   CLIENT_SECRETS_FILE))

# An OAuth 2 access scope that allows for full read/write access
YOUTUBE_SCOPE = "https://www.googleapis.com/auth/youtube"
YOUTUBE_API_SERVICE_NAME = "youtube"
YOUTUBE_API_VERSION = "v3"

def get_authenticated_service():
  flow = flow_from_clientsecrets(CLIENT_SECRETS_FILE, scope=YOUTUBE_SCOPE,
    message=MISSING_CLIENT_SECRETS_MESSAGE)

  storage = Storage("%s-oauth2.json" % sys.argv[0])
  credentials = storage.get()

  if credentials is None or credentials.invalid:
    credentials = run(flow, storage)

  return build(YOUTUBE_API_SERVICE_NAME, YOUTUBE_API_VERSION,
    http=credentials.authorize(httplib2.Http()))

def add_featured_video(options):
  youtube = get_authenticated_service()

  add_video_request = youtube.channels().update(
    part="invideoPromotion",
    # Test different payloads in the API explorer:
    #    https://developers.google.com/youtube/v3/docs/channels/update#try-it
    body={
      "invideoPromotion": {
        "position": {
          "cornerPosition": options.position,
          "type": "corner"
      },
      "items": [{
        "type": "video",
        "videoId": options.video_id
      }],
      "timing": {
        "offsetMs": options.offset_ms,
        "type": options.offset_type
      }
    },
    "id": options.channel_id
  })

  add_video_response = add_video_request.execute()
  print "Added featured video %s to channel %s." % (
      add_video_response["invideoPromotion"]["items"][0]["videoId"],
      add_video_response["id"])

# If offsetMs or position are not valid, the API will throw an error
VALID_OFFSET_TYPES = ("offsetFromEnd", "offsetFromStart",)
VALID_POSITIONS = ("topLeft", "topRight", "bottomLeft", "bottomRight",)

if __name__ == '__main__':
  parser = OptionParser()
  parser.add_option("--channel_id", dest="channel_id", help="Channel ID of the channel to add a featured video")
  parser.add_option("--video_id", dest="video_id", help="Video ID to feature on your channel")
  parser.add_option("--position", dest="position",
    help="Position to show promotion. Options are: %s" % ", ".join(VALID_POSITIONS),
    default="bottomLeft")
  parser.add_option("--offset_ms", dest="offset_ms",
    help="Offset in milliseconds to show video. Default is 10000, or 10 seconds",
    default="10000")
  parser.add_option("--offset_type", dest="offset_type",
    help="Describes whether the offset is from the beginning or end of video playback."
      + " Valid options are: %s" % ",".join(VALID_OFFSET_TYPES),
    default="offsetFromEnd")
  (options, args) = parser.parse_args()

  # Require a channel ID and video ID
  if options.channel_id is None:
    exit("Please specify a valid channel ID using the --channel_id parameter.")
  elif options.video_id is None:
    exit("Please specify a valid video ID to feature using the --video_id parameter.")
  # Validate offset type and position parameters
  if options.offset_type not in VALID_OFFSET_TYPES:
    exit("offset_type must be one of: %s" % ",".join(VALID_OFFSET_TYPES))
  if options.position not in VALID_POSITIONS:
    exit("position must be one of: %s" % ", ".join(VALID_POSITIONS))
  else:
    add_featured_video(options)
